src/coprocessor/dag/batch/executors/stream_aggr_executor.rs

```diff
@@ -0,0 +1,515 @@
+                let group_column = group_by_result.vector_value().unwrap();
+                group_key.push(group_column.get_unchecked(row_index));
+            }
+            match self.keys.rchunks_exact(group_by_len).next() {
+                Some(current_key) if &group_key[..] == current_key => {
+                    group_key.clear();
+                }
+                _ => {
+                    // Update the complete group
+                    if row_index > 0 {
+                        update_current_states(
+                            context,
+                            &mut self.states,
+                            aggr_fn_len,
+                            &aggr_expr_results,
+                            group_start_row,
+                            row_index,
+                        )?;
+                    }
+
+                    // create a new group
+                    group_start_row = row_index;
+                    self.keys.extend(group_key.drain(..).map(Into::into));
+                    for aggr_fn in &entities.each_aggr_fn {
+                        self.states.push(aggr_fn.create_state());
+                    }
+                }
+            }
+        }
+        // Update the current group with the remaining data
+        update_current_states(
+            context,
+            &mut self.states,
+            aggr_fn_len,
+            &aggr_expr_results,
+            group_start_row,
+            rows_len,
+        )?;
+
+        Ok(())
+    }
+
+    /// Note that the partial group is in count.
+    #[inline]
+    fn groups_len(&self) -> usize {
+        self.keys
+            .len()
+            .checked_div(self.group_by_exps.len())
+            .unwrap_or(0)
+    }
+
+    #[inline]
+    fn iterate_each_group_for_aggregation(
+        &mut self,
+        entities: &mut Entities<Src>,
+        src_is_drained: bool,
+        mut iteratee: impl FnMut(&mut Entities<Src>, &[Box<dyn AggrFunctionState>]) -> Result<()>,
+    ) -> Result<Vec<LazyBatchColumn>> {
+        let number_of_groups = if src_is_drained {
+            AggregationExecutorImpl::<Src>::groups_len(self)
+        } else {
+            // don't include the partial group
+            AggregationExecutorImpl::<Src>::groups_len(self) - 1
+        };
+
+        let group_by_exps_len = self.group_by_exps.len();
+        let mut group_by_columns: Vec<_> = self
+            .group_by_exps_types
+            .iter()
+            .map(|tp| LazyBatchColumn::decoded_with_capacity_and_tp(number_of_groups, *tp))
+            .collect();
+        let aggr_fns_len = entities.each_aggr_fn.len();
+
+        // key and state ranges of complete groups
+        let keys_range = ..number_of_groups * group_by_exps_len;
+        let states_range = ..number_of_groups * aggr_fns_len;
+
+        for states in self.states[states_range].chunks_exact(aggr_fns_len) {
+            iteratee(entities, states)?;
+        }
```

Comment:
I will feel safe to assert remainder is none.

Reply:
`for` consumes the `ChunksExact`. So I think the code will become a bit uglier if we assert the remainder is empty...

Reply:
or maybe we can check it beforehand?

Reply:
Not necessary. It panics at `self.states[states_range]` if index out of range. And because states_range length is exactly number_of_groups * aggr_fns_len, the chunks won't have any remainder.

Reply:
Yeah... so I suggested an assertion instead of some other kind of checking. Anyway, assertion is optional, only for defending purpose.